File: SinglyLinkedList.cpp
#include<stdio.h>
#include<stdlib.h>
#include<stack>
using namespace std;

struct Node{
	int data;
	Node* next;
};

struct Node* head = NULL;
struct Node* tail = NULL;
int size = 0;

Node* createNewNode(int data){
	Node* temp = new Node();
	temp->next = NULL;
	temp->data = data;
	return temp;
}

void insertNode(int data){
	Node* node = createNewNode(data);
	if(head == NULL){
		head = node;
	}else{
		tail->next = node;
	}
	size++;
	tail = node;
}

void insertAtHead(int data){
	Node* node = createNewNode(data);
	if(head == NULL){
		head = node;
	}else{
		node->next = head;
		head = node;
	}
	size++;
}

void insertAt(int data, int n){
	Node* temp = head;
	if(n == size){
		insertNode(data);
		return;
	}
	if(n == 1){
		insertAtHead(data);
		return;
	}
	for(int i = 0; i < n-2; i++){ // Go to (n-1)th node
		temp = temp->next;
	}
	Node* node = createNewNode(data);
	node->next = temp->next; // Build link to nth to n+1 node
	temp->next = node; // Build link n-1 node to nth node
	size++;
}

void print(){
	Node* temp = head;
	if(head == NULL){
		return;
	}
	while(temp != NULL){
		printf("%d ",temp->data);
		temp = temp->next;
	}
	printf("\n");
}

void del(int n){
	Node* temp1 = head;
	if(n == 1){
		head = head->next;
		free(temp1);
		return;
	}
	// Traversing to n-1 node
	for(int i = 0; i < n-2; i++){
		temp1 = temp1->next;
	}
	Node* temp2 = temp1->next; // temp2 points to nth node
	temp1->next = temp2->next; // Go to (n+1)th node
	free(temp2);
	if(n == size){
		tail = temp1;
	}
	size--;
}

void reversePrint(Node* head){
	if(head == NULL){
		return;
	}else{
		reversePrint(head->next);
	}
	printf("%d ",head->data);
}

void reverse(){
	stack<Node*> S;
	Node* temp = head;
	while(temp != NULL){
		S.push(temp);
		temp = temp->next;
	}
	temp = S.top();
	head = temp;
	S.pop();
	while(!S.empty()){
		temp->next = S.top();
		temp = temp->next;
		S.pop();
	}
	temp->next = NULL;
	tail = temp;
}

int main(){
	insertNode(3);
	insertNode(4);
	insertNode(5);
	insertNode(6);
	reverse();
	print();
	insertAt(7,4);
	print();
	return 0;
}

File: SinglyLinkedList/src/SinglyLinkedList.java
import java.util.Stack;

public class SinglyLinkedList{
    static class Node{
        public int data;
        public Node next;
        public Node(){}
        public Node(int data){
            this.data = data;
            this.next = null;
        }
    }

    public Node head;
    public Node tail;
    public int size;

    public SinglyLinkedList(){
        this.head = null;
        this.tail = null;
    }

    public void insertNode(int data){
        Node node = new Node(data);
        if(head == null){
            head = node;
        }else{
            tail.next = node;
        }
        tail = node;
        size++;
    }

    public void insertNode(int data, int n){
        Node temp1 = new Node(data);
        if(n == 1){
            temp1.next = head;
            head = temp1;
            tail = temp1;
            size++;
            return;
        }

        if(n == size){
            this.insertNode(data);
            return;
        }

        Node temp2 = head;
        for(int i = 0; i < n - 2; i++){
            temp2 = temp2.next;
        }
        temp1.next = temp2.next;
        temp2.next = temp1;

    }

    public void delete(int n){
        Node temp1 = head;
        if(n == 1){
            head = temp1.next;
            temp1 = null;
            return;
        }
        for(int i = 0; i < n-2; i++){
            temp1 = temp1.next;
        }
        // temp1 is pointing to n-1 node
        Node temp2 = temp1.next; // nth Node

        if(n == size){
            tail = temp1;
        }

        temp1.next = temp2.next; // (n+1)th Node
        temp2 = null; // free temp2
        size--;
    }

    // I love you 3000, Ngoc Anh, love you,

    public void print(){
        Node p = head;
        if(p == null){
            return;
        }
        while(p != null){
            System.out.print(p.data + " ");
            p = p.next;
        }
        System.out.println();
    }

    public void reverse(){
        Node current, prev, next;
        current = head;
        tail = current;
        prev = null;
        while(current != null){
            next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }
        head = prev;
    }

    public void reverseUsingStack(){
        if(head == null)
            return;
        Stack<Node> stack = new Stack<>();
        Node temp = head;
        while(temp != null){
            stack.push(temp);
            temp = temp.next;
        }

        temp = stack.pop();
        head = temp;
        while(!stack.isEmpty()){
            temp.next = stack.pop();
            temp = temp.next;
        }
        temp.next = null;
        this.tail = temp;
    }

}